keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -326,75 +284,105 @@ spec:
+- Bump severity of the alert to Warning.
+- Announce the behavior change and deprecate the old behavior.
+
+If Phase 1 shows that too many applications would be broken, then go GA only with Phase 1, i.e. `ReadWriteOncePod` PVCs.
+Even that will help users to avoid recursive relabeling of volumes if their application can use `ReadWriteOncePod` PVCs.
+
 ### Test Plan
 
 * Unit tests:
-   * API validation (all permutations missing / present PodSecurityPolicy.SELinuxOptions & SELinuxRelabelPolicy & container.SecurityPolicy.SELinuxOptions)
    * Passing mount options from kubelet to volume plugins.
 * E2e tests:
-   * Check no recursive `chcon` is done on a volume when not needed /
-   * Check recursive `chcon` is done on a volume when needed (with a matrix of SELinuxOptions / SELinuxRelabelPolicy).
+   * Check no recursive `chcon` is done on a volume when not needed.
+   * Check recursive `chcon` is done on a volume when needed.
+   * Check that proper metric + alert is emitted when kubelet can't start two pods with different SELinux contexts using the same volume on the same node.
+   * These tests might use only CSI volumes, GCE PD in-tree volume plugin that we use for e2e tests might be already migrated to CSI by that time. 
 * Prepare e2e job that runs with SELinux in Enforcing mode!
 
 ### Graduation Criteria
 
-* Alpha:
- * Provided all tests defined above are passing and gated by the feature gate `SELinuxRelabelPolicy` and set to a default of `false`.
- * Documentation exists.
-* Beta: with discussions in SIG-Storage regarding success of deployments. A metric will be added to report time taken to perform a volume ownership change. Feature gate `ConfigurableFSGroupPolicy` is `true`.
-* GA: all known issues fixed.
+* Alpha of Phase 1:
+  * Provided all tests defined above are passing and gated by the feature gate `SELinuxMountReadWriteOncePod` and set to a default of `false`.
+  * Documentation exists.
+* Beta of Phase 1: 
+  * The feature gate is `true` by default.
+* Evaluation:
+  * During the next release after Phase 1 is beta (= the feature is enabled by default), collect reports from users about possible breakage.
+  * KEP author has access to usage data from OpenShift, a Kubernetes distro that runs with SELinux in enforcing mode.
+* Alpha of Phase 2:
+  * Only if nr. of broken apps is low!
+    * To be discussed in sig-storage and sig-arch?. 
+  * Publish deprecation note about changed behavior.
+  * Implement Phase 2 **with a separate alpha feature gate `SELinuxMount`**.
+* GA: all known issues fixed + deprecation period is over. Otherwise, we will GA Phase 1 only.
 
 ### Upgrade / Downgrade Strategy
 
-`SELinuxRelabelPolicy` becomes "invisible" or dropped in an downgraded cluster. Container runtime will get ":Z" on volumes and they will do slow recursive chown as they do today.
+N/A. This feature affects only mounts. It does not depend on version of Kubernetes on other nodes or in the control plane.
```

Comment:
Do you see any issue if a driver is upgraded to support this new mode while Pods are still mounted the old way? Should we also have a test for driver upgrade?

Reply:
Good point, some drivers may have issues:

1. Kubelet or CSIDriver object does not support mounting with `-o context` -> NodeStage is called without it.
2. Kubelet is restarted or CSIDriver object is updated to announce `-o context` support -> NodePublish is called with it. This may fail, because the volume was mounted during NodeStage without `-o context` and it can't be changed afterwards.

Similarly, idempotency may be more complicated. NodePublish without `-o context` may be followed by NodePublish with it, as CSIDriver got updated and kubelet restarted and kubelet is not sure about the volume status and tries to fix its state of world.

We can't keep status how the volume was staged in kubelet memory because of restarts. We can't assume that a CSI driver remembers the mount options from NodeStage and applies the same options in NodePublish. Can we keep the options we used for NodeStage in the volume json in the global dir? Similarly, we need to store if the initial NodePublish was called with SELinux options or not, for volumes that don't implement NodeStage.

And yes, this will need tests and they will be very interesting.

Reply:
> This may fail, because the volume was mounted during NodeStage without -o context and it can't be changed afterwards.

I wonder if it helps to have explicit selinux mount rule in CSI driver. ie. - `SELinuxMountPolicy: OnStage|OnPublish|None` . This way kubelet does not try to guess when to apply selinux mount options.

> Similarly, idempotency may be more complicated. NodePublish without -o context may be followed by NodePublish with it, as CSIDriver got updated and kubelet restarted and kubelet is not sure about the volume status and tries to fix its state of world.

Assuming `CSIDriver.Spec.SELinuxMountPolicy: OnPublish` - the "correction" should happen only when volume is remounted right? So - we should be fine?

Reply:
> I wonder if it helps to have explicit selinux mount rule in CSI driver. ie. - SELinuxMountPolicy: OnStage|OnPublish|None . This way kubelet does not try to guess when to apply selinux mount options.

Kubelet passes the same mount options both to NodeStage and NodePublish. IMO we can't expect that only one of the calls actually uses them. We would need some `OnBoth` option and then kubelet must make sure it does not change the options between NodeStage and NodePublish. And therefore I think we can implement this behavior right away, pass the mount options to NodeStage and NodePublish and make sure they don't change in between.

> Assuming CSIDriver.Spec.SELinuxMountPolicy: OnPublish - the "correction" should happen only when volume is remounted right? So - we should be fine?

I am not sure about kubelet restart - can kubelet call NodePublish to resolve uncertain mount after reconstruction? Kubelet can't expect that a volume is a mount point. At least [this](https://github.com/kubernetes/kubernetes/blob/f49a15ae50196fc4d46d70f86de4d15cbc2405c6/pkg/kubelet/volumemanager/reconciler/reconciler.go#L393-L397) says that SetUp will be called after reconstruction.

Reply:
> We would need some OnBoth option and then kubelet must make sure it does not change the options between NodeStage and NodePublish. And therefore I think we can implement this behavior right away, pass the mount options to NodeStage and NodePublish and make sure they don't change in between.

Usually you are right. But why does it matter for SELinux options? SELinux mount options can be applied only either during stage or publish - not both.  I was trying to avoid a scenario where node stage was *not* mounted using selinux context and then node-publish tries to use it. 

> I am not sure about kubelet restart - can kubelet call NodePublish to resolve uncertain mount after reconstruction? Kubelet can't expect that a volume is a mount point. At least this says that SetUp will be called after reconstruction.

Looking at the code - `NodePublish` can get called after reconstruction. So - we may try to mount a volume with selinux context when we previously did not.

Reply:
BTW we have/had same problem for fsgroup on mount.

Reply:
I added a note about tracking SELinux context during volume reconstruction, kubelet must reconstruct what SELinux label the previous kubelet use. For in-tree volumes we can get it from mount table, for CSI we can have it in the json.